### PR TITLE
[coloattention] coloattention support flash attention 2

### PR DESCRIPTION
**Pull request by flybird1111**
## 📌 Checklist before creating the PR

- [ ] I have created an issue for this PR for traceability
- [x] The title follows the standard format: `[doc/gemini/tensor/...]: A concise description`
- [x] I have added relevant tags if possible for us to better distinguish different PRs

## 🚨 Issue number

> Link this PR to your issue with words like fixed to automatically close the linked issue upon merge
>
> e.g. `fixed #1234`, `closed #1234`, `resolved #1234`

#4322 

## 📝 What does this PR do?

> Summarize your work here.
> if you have any plots/diagrams/screenshots/tables, please attach them here.

coloattention support flash attention 2

## 💥 Checklist before requesting a review

- [x] I have linked my PR to an issue ([instruction](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue))
- [ ] My issue clearly describes the problem/feature/proposal, with diagrams/charts/table/code if possible
- [x] I have performed a self-review of my code
- [x] I have added thorough tests.
- [ ] I have added docstrings for all the functions/methods I implemented

## ⭐️ Do you enjoy contributing to Colossal-AI?

- [ ] 🌝 Yes, I do.
- [ ] 🌚 No, I don't.

Tell us more if you don't enjoy contributing to Colossal-AI.

### CODE REVIEW

File: colossalai/kernel/cuda_native/fmha/fmha.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                                                   seq_lens_info_q.indices).unbind(dim=2)
+                    repad_indices = seq_lens_info_q.indices
+            else:
+                seq_lens_info_q = SeqLensInfo.materialize(batch_size, tgt_len, dtype=torch.int32, device=query.device)
+                seq_lens_info_kv = self.get_seq_info_from_mask(attn_mask, query.device)
+                if batch_size > 1:
+                    query = rearrange(query, "b s ... -> c (b s) ...", c=1)
+                    key, value = self.unpad(torch.stack([query, key, value], dim=2),
+                                            seq_lens_info_kv.indices).unbind(dim=2)
+                    repad_indices = seq_lens_info_kv.indices
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/fmha/fmha.py:87]: Isn't it supposed to be indices of q?

File: colossalai/kernel/cuda_native/fmha/fmha.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+                key: torch.Tensor,
+                value: torch.Tensor,
+                attn_mask: Optional[torch.Tensor] = None,
+                attn_mask_type: Optional[AttnMaskType] = None,
+                bias: Optional[torch.Tensor] = None):
+
+        do = None
+        if HAS_FLASH_ATTN and query.dtype in [torch.float16, torch.bfloat16] and bias == None:
+            from .flash_attn_2 import flash_attention
+            do = flash_attention
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/fmha/fmha.py:55]: how about `attention` instead of `do`

File: colossalai/kernel/cuda_native/fmha/fmha.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        if scale is not None:
+            self.scale = scale
+        else:
+            self.scale = 1 / math.sqrt(embed_dim // num_heads)
+        self.dropout = dropout
+
+        if not HAS_MEM_EFF_ATTN and not HAS_FLASH_ATTN:
+            raise Exception("flash attention can not support!")
+
+    def get_seq_info_from_mask(self, attn_mask: torch.Tensor, device="cuda"):
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/fmha/fmha.py:29]: do this calculation in SeqLenInfo.materialize, then we don't need this method in ColoAttention any more

File: colossalai/kernel/cuda_native/fmha/utils.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        indices, = ctx.saved_tensors
+        # [b*s, ...]
+        grad_output = rearrange(grad_output, 'b s ... -> (b s) ...')
+        grad = grad_output[indices]
+        # [1, ntokens, ...]
+        return grad.unsqueeze(0), None, None, None
+
+
+@dataclass
+class SeqLensInfo:
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/fmha/utils.py:66]: Use `SeqLenInfo`

File: colossalai/kernel/cuda_native/fmha/utils.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    seqlens: Iterable[int] = None
+    indices: torch.Tensor = None
+    max_seqlen: int = None
+    cu_seqlens: torch.Tensor = None
+
+    @staticmethod
+    def materialize(batch_size, tgt_len, dtype=torch.int32, device='cuda'):
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/fmha/utils.py:73]: As commented at `get_seq_info_from_mask`

File: colossalai/kernel/cuda_native/fmha/flash_attn_2.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+    from .utils import SeqLensInfo
+
+    def flash_attention(q,
+                        k,
+                        v,
+                        seq_lens_info_q: SeqLensInfo,
+                        seq_lens_info_kv: SeqLensInfo,
+                        dropout_p=0.,
+                        sm_scale=None,
+                        attn_mask_type=None):
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/fmha/flash_attn_2.py:42]: It may be more clear if we check attn mask type in ColoAttention then pass `pad` and `causal` to this interface

File: colossalai/kernel/cuda_native/fmha/mem_eff_attn.py [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    )
+
+    from ..scaled_softmax import AttnMaskType
+    from .utils import SeqLensInfo
+
+    allow_alibi = True
+    for op in MemoryEfficientAttentionCutlassOp:
+        allow_alibi = allow_alibi & (LowerTriangularMaskWithTensorBias in op.SUPPORTED_ATTN_BIAS_TYPES)
+
+    def mem_eff_attention(query: torch.Tensor,
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/fmha/mem_eff_attn.py:34]: Make the parameters consistent with flash version

**Review by kurisusnowdeng — CHANGES_REQUESTED** [review]

**Review by kurisusnowdeng — CHANGES_REQUESTED** [review]

File: colossalai/kernel/cuda_native/mha/mem_eff_attn.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+    def mem_eff_attention(q: torch.Tensor,
+                          k: torch.Tensor,
+                          v: torch.Tensor,
+                          seq_lens_info_q: SeqLenInfo,
+                          seq_lens_info_kv: SeqLenInfo,
+                          dropout_p: float = 0.,
+                          sm_scale: float = None,
+                          causal: bool = False,
+                          bias: Optional[torch.Tensor] = None,
+                          **kwargs):
```
[the remark this answers is not in the document]
Comment (flybird1111) [reply colossalai/kernel/cuda_native/mha/mem_eff_attn.py]: updated.

File: colossalai/kernel/cuda_native/mha/flash_attn_2.py [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+except ImportError:
+    warnings.warn('please install flash_attn from https://github.com/HazyResearch/flash-attention')
+    HAS_FLASH_ATTN = False
+
+if HAS_FLASH_ATTN:
+    from einops import rearrange
+
+    from .utils import SeqLenInfo
+
+    def flash_attention(q: torch.Tensor,
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/mha/flash_attn_2.py:35]: use this format 
```
flash_attention(q: torch.Tensor,
                k: torch.Tensor,
                v: torch.Tensor,
                seq_len_info_q: SeqLenInfo = None,
                seq_len_info_kv: SeqLenInfo = None,
                bias: Optional[torch.Tensor] = None,
                dropout_p: float = 0.,
                scale: float = None,
                causal: bool = False,
                padded: bool = False)
```

File: colossalai/kernel/cuda_native/fmha/fmha.py (hunk 1 above)
Comment (flybird1111) [reply colossalai/kernel/cuda_native/fmha/fmha.py]: > Isn't it supposed to be indices of q?

updated.

File: colossalai/kernel/cuda_native/fmha/fmha.py (hunk 2 above)
Comment (flybird1111) [reply colossalai/kernel/cuda_native/fmha/fmha.py]: > how about `attn` instead of `do`

updated.

File: colossalai/kernel/cuda_native/fmha/fmha.py (hunk 3 above)
Comment (flybird1111) [reply colossalai/kernel/cuda_native/fmha/fmha.py]: > do this calculation in SeqLenInfo.materialize, then we don't need this method in ColoAttention any more

updated.

File: colossalai/kernel/cuda_native/fmha/flash_attn_2.py (hunk 6 above)
Comment (flybird1111) [reply colossalai/kernel/cuda_native/fmha/flash_attn_2.py]: > It may be more clear if we check attn mask type in ColoAttention then pass `pad` and `causal` to this interface

updated.

File: colossalai/kernel/cuda_native/mha/utils.py [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+@dataclass
+class SeqLenInfo:
+    seqlens: Iterable[int] = None
+    indices: torch.Tensor = None
+    max_seqlen: int = None
+    cu_seqlens: torch.Tensor = None
+
+    @staticmethod
+    def materialize(attn_mask: torch.Tensor, from_mask=False, device='cuda'):
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/mha/utils.py:73]: make this like:
```
@staticmethod
def materialize(attn_mask: torch.Tensor = None, size: Tuple[int] = None, device=get_current_device()):
    if attn_mask is None:
        batch_size, tgt_len = size[0], size[1]
        ...
    else:
        do from mask ...
```
Note:

1. get current device instead of cuda
2. make `seqlens` to list at the end, so that we don't need to make it tensor again for cu_seqlens

File: colossalai/kernel/cuda_native/mha/mha.py [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+        pad = False
+        causal = False
+        if attn_mask_type and attn_mask_type.value % 2 == 1:
+            pad = True
+        if attn_mask_type == AttnMaskType.causal:
+            causal = True
+
+        batch_size, tgt_len = query.shape[0], query.shape[1]
+        # unpad
+        seq_lens_info_q = SeqLenInfo()
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/mha/mha.py:62]: Initially they should be None

File: colossalai/kernel/cuda_native/mha/flash_attn_2.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+            batch_size: int.
+            seq_len: int.
+            dropout_p: float. Dropout probability.
+            sm_scale: float. The scaling of QK^T before applying softmax.
+                Default to 1 / sqrt(headdim).
+            causal: bool. Whether to apply causal attention mask (e.g., for auto-regressive modeling).
+        Return:
+            attn_out: (batch, q_seqlen, nheads, headdim).
+        """
+        if seq_lens_info_q.cu_seqlens == None:
```
Comment (kurisusnowdeng) [inline colossalai/kernel/cuda_native/mha/flash_attn_2.py:59]: just check `pad`

**Review by kurisusnowdeng — CHANGES_REQUESTED** [review]

**Review by kurisusnowdeng — APPROVED** [review]

**Comment by flybird1111** [comment]
image
All tests have passed.

**Comment by flybird1111** [comment]
image
All tests have passed.

**Comment by flybird1111** [comment]
image
All tests have passed.
